### ISSUE TITLE
VLLM Multi-Lora with embed_tokens and lm_head in adapter weights 

### ISSUE THREAD

**Issue by germanjke** [vllm-project/vllm#2816]
Hi there!

I've encountered an issue with the `adatpter_model.safetensors` in my project, and I'm seeking guidance on how to handle `lm_head` and `embed_tokens` within the specified modules. Here's the current state:

```
['base_model.model.lm_head.weight',
 'base_model.model.model.embed_tokens.weight',
 'base_model.model.model.layers.0.self_attn.k_proj.lora_A.weight',
 'base_model.model.model.layers.0.self_attn.k_proj.lora_B.weight',
 'base_model.model.model.layers.0.self_attn.o_proj.lora_A.weight',
 'base_model.model.model.layers.0.self_attn.o_proj.lora_B.weight',
 ...]
 ```
 
Firstly, in [vllm/lora/utils.py](https://github.com/vllm-project/vllm/blob/0e163fce18594c7e29dc5a143dd6b33d213fcbf3/vllm/lora/utils.py#L33), modifications are needed to work with `lm_head` and `embed_tokens`. My question is, what should be the appropriate name for the new module? For instance, from `base_model.model.model.layers.0.self_attn.k_proj.lora_A.weight`, we extract `model.layers.0.self_attn.k_proj`. What should I use for `base_model.model.lm_head.weight` and `base_model.model.model.embed_tokens.weight`? Are `model.lm_head` and `model.embed_tokens` suitable, or do you have alternative suggestions?

Additionally, in [vllm/lora/models.py](https://github.com/vllm-project/vllm/blob/931746bc6d7c1c0ab40b2c4f58b51b855f0b2c94/vllm/lora/models.py#L176), there is a check for `if embedding:` followed by placement in `embedding_tensor`, which is optional in [LoRALayerWeights](https://github.com/vllm-project/vllm/blob/f0d4e145575bf6fb96c141d776ce92c9bfc79c49/vllm/lora/lora.py#L17C9-L17C26). However, `lora_a` and `lora_b` are not optional there. When using `activate_lora` with [module.set_lora](https://github.com/vllm-project/vllm/blob/f0d4e145575bf6fb96c141d776ce92c9bfc79c49/vllm/lora/models.py#L370C24-L370C32), should we set lora to `lm_head` and `embed_tokens` if they are part of adapter weights? Is it `set_lora` and not `reset_lora`?

Finally, there's an error in [set_lora](https://github.com/vllm-project/vllm/blob/main/vllm/lora/layers.py#L226) when `lora_a is None`, causing issues with `.shape`. @Yard1, do you have any insights or suggestions on working with loras involving special tokens? Classic loras seem to work well with multi-lora setups.

Thank you for your assistance!

**Comment by Yard1** [comment]
Hi @germanjke, is it possible for you to share the LoRA for debugging?

**Comment by germanjke** [comment]
Hi @Yard1!

I hope this message finds you well. After conducting some research, we have identified a potential issue in the code logic related to the retraining of LoRA model embeddings. Currently, it seems that the code supports the retraining of only a few new subtokens (e.g., 5), rather than the entire expanded `embed_tokens` comprising both the old tokens (e.g., 32000) and the new ones (32005 in sum).

The logic for this limitation can be observed in [this section of the code](https://github.com/vllm-project/vllm/blob/c81dddb45c71e630b907f9d84686ecd73b4105c7/vllm/engine/arg_utils.py#L43), where the parameter extra_vocab_size is used.

The current situation leads to an error in our LoRA model, as evidenced by the following traceback:

```
File "/opt/tritonserver/vllm-src/vllm/lora/worker_manager.py", line 236, in add_lora
    self._lora_manager.activate_lora(lora_request.lora_int_id)
  File "/opt/tritonserver/vllm-src/vllm/lora/models.py", line 624, in activate_lora
    result = super().activate_lora(lora_id)
  File "/opt/tritonserver/vllm-src/vllm/lora/models.py", line 372, in activate_lora
    module.set_lora(index, module_lora.lora_a, module_lora.lora_b,
  File "/opt/tritonserver/vllm-src/vllm/lora/layers.py", line 237, in set_lora
    self.embeddings_tensors[
RuntimeError: The size of tensor a (256) must match the size of tensor b (32005) at non-singleton dimension 0
```

To address this issue, we are considering training a new LoRA model where only the new tokens (e.g., 5) are included in the `embed_tokens`. This approach aims to avoid making changes to the backend compile code, which could be time-consuming and may not be the best practice.

Before proceeding, we want to confirm our understanding of your logic. It seems that you prefer not to support LoRA models with full embeddings retraining at the lora-stage, including all the old tokens from the base model. Could you please confirm if our interpretation is correct?

Thank you for your time and assistance.

**Comment by Yard1** [comment]
That's correct, the base tokens should be modified through `embed_tokens.lora_a` and `embed_tokens.lora_b`, and the weights for new tokens should be included in `new_embeddings.safetensors`.

**Comment by tensimixt** [comment]
@Yard1 does this approach also apply to 
` model.layers.0.input_layernorm.lora_A.weight ` 
` model.layers.0.input_layernorm.lora_B.weight `

As in the need to create new_embeddings.safetensors which will include input_layernorm?

**Comment by prashantgupta24** [comment]
> Hi @germanjke, is it possible for you to share the LoRA for debugging?
> 
> For both `lm_head` and `embed_tokens` we expect normal LoRAs. However, if you have extra vocab not present in the base model, you should provide separate `new_embeddings.safetensors` file, which will contain weights to be used for additional vocab. Therefore, this is the expected structure:
> 
> `adapter_model.safetensors`:
> 
> ```
> ['base_model.model.lm_head.lora_A.weight',
>  'base_model.model.lm_head.lora_B.weight',
>  'base_model.model.model.embed_tokens.lora_A.weight',
>  'base_model.model.model.embed_tokens.lora_B.weight',
>  ...]
> ```
> 
> `new_embeddings.safetensors`:
> 
> ```
> ['lm_head.weight',
>  'embed_tokens.weight',
> ]
> ```

@Yard1 , looking at the `new_embeddings.safetensors` file you have [here](https://github.com/vllm-project/vllm/blob/v0.6.0/tests/lora/test_lora_manager.py#L33) and the `EMBEDDING_MODULES` declared as [this](https://github.com/vllm-project/vllm/blob/v0.6.0/tests/lora/test_lora_manager.py#L21-L23): 

```
EMBEDDING_MODULES = {
    "embed_tokens": "input_embeddings",
    "lm_head": "output_embeddings",
}
```
I see:
```
>>> new_embed = "cache/models--yard1--llama-2-7b-sql-lora-test/snapshots/0dfa347e8877a4d4ed19ee56c140fa518470028c/new_embeddings.safetensors"
>>> f = safetensors.safe_open(new_embed, framework="pt")
>>> mods = [mod for mod in f.keys()]
>>> mods
['input_embeddings', 'output_embeddings']
```

Should the content of the `new_embeddings.safetensors` contain `output_embeddings` instead of `lm_head.weight` then?
